pygeos/creation.py

```diff
@@ -125,45 +125,50 @@ def linearrings(coords, y=None, z=None, indices=None, **kwargs):
 
 
 @multithreading_enabled
-def polygons(shells, holes=None, indices=None, **kwargs):
+def polygons(geometries, holes=None, indices=None, **kwargs):
     """Create an array of polygons.
 
     Parameters
     ----------
-    shell : array_like
-        An array of linearrings that constitute the out shell of the polygons.
-        Coordinates can also be passed, see linearrings.
+    geometries : array_like
+        An array of linearrings or coordinates (see linearrings).
+        If ``indices`` are given, the geometries include both the outer shell
+        and the holes. If not, the geometries only include the shells and the
+        ``holes`` argument should be used to create polygons with holes.
     holes : array_like or None
         An array of lists of linearrings that constitute holes for each shell.
+        Not to be used in combination with ``indices``.
     indices : array_like or None
-        Indices into the shells array where input holes belong. If
-        provided, shells, holes, and indices should all be 1D and the size
-        of holes must equal the size of indices.
+        Indices into the target array where input geometries belong. If
+        provided, the holes are expected to be present inside ``geometries``.
+        Both geometries and indices should be 1D and have matching  sizes.
     **kwargs
         For other keyword-only arguments, see the
         `NumPy ufunc docs <https://numpy.org/doc/stable/reference/ufuncs.html#ufuncs-kwargs>`_.
         Ignored if ``indices`` is provided.
     """
-    shells = np.asarray(shells)
-    if not isinstance(shells, Geometry) and np.issubdtype(shells.dtype, np.number):
-        shells = linearrings(shells)
+    geometries = np.asarray(geometries)
+    if not isinstance(geometries, Geometry) and np.issubdtype(
+        geometries.dtype, np.number
+    ):
+        geometries = linearrings(geometries)
+
+    if indices is not None:
+        if holes is not None:
+            raise TypeError("Indices provided with a holes array.")
```

Comment:
```suggestion
            raise TypeError("Cannot specify separate holes array when using indices.")
```

(might be slightly clearer message)